// telegram/tl/custom/dialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dialog = void 0;
const api_1 = require("../api");
const Utils_1 = require("../../Utils");
const draft_1 = require("./draft");
class Dialog {
    constructor(client, dialog, entities, message) {
        var _a;
        this._client = client;
        this.dialog = dialog;
        this.pinned = !!dialog.pinned;
        this.folderId = dialog.folderId;
        this.archived = dialog.folderId != undefined;
        this.message = message;
        this.date = (_a = this.message) === null || _a === void 0 ? void 0 : _a.date;
        this.entity = entities.get(Utils_1.getPeerId(dialog.peer));
        this.inputEntity = Utils_1.getInputPeer(this.entity);
        if (this.entity) {
            this.id = Utils_1.getPeerId(this.entity); // ^ May be InputPeerSelf();
            this.name = this.title = Utils_1.getDisplayName(this.entity);
        }
        this.unreadCount = dialog.unreadCount;
        this.unreadMentionsCount = dialog.unreadMentionsCount;
        if (!this.entity) {
            throw new Error("Entity not found for dialog");
        }
        this.draft = new draft_1.Draft(client, this.entity, this.dialog.draft);
        this.isUser = this.entity instanceof api_1.Api.User;
        this.isGroup = !!((this.entity instanceof api_1.Api.Chat &&
            this.entity instanceof api_1.Api.ChatForbidden) ||
            (this.entity instanceof api_1.Api.Channel && this.entity.megagroup));
        this.isChannel = this.entity instanceof api_1.Api.Channel;
    }
}
exports.Dialog = Dialog;
