// telegram/tl/custom/forward.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Forward = void 0;
const chatGetter_1 = require("./chatGetter");
const senderGetter_1 = require("./senderGetter");
const Helpers_1 = require("../../Helpers");
const Utils_1 = require("../../Utils");
class Forward {
    constructor(client, original, entities) {
        // Copy all objects here. probably need a better way tho. PRs are welcome
        Object.assign(this, original);
        this.originalFwd = original;
        let senderId = undefined;
        let sender = undefined;
        let inputSender = undefined;
        let peer = undefined;
        let chat = undefined;
        let inputChat = undefined;
        if (original.fromId) {
            const ty = Helpers_1._entityType(original.fromId);
            if (ty === Helpers_1._EntityType.USER) {
                senderId = Utils_1.getPeerId(original.fromId);
                [sender, inputSender] = Utils_1._getEntityPair(senderId, entities, client._entityCache);
            }
            else if (ty === Helpers_1._EntityType.CHANNEL || ty === Helpers_1._EntityType.CHAT) {
                peer = original.fromId;
                [chat, inputChat] = Utils_1._getEntityPair(Utils_1.getPeerId(peer), entities, client._entityCache);
            }
        }
        chatGetter_1.ChatGetter.initClass(this, {
            chatPeer: peer,
            inputChat: inputChat,
        });
        senderGetter_1.SenderGetter.initClass(this, {
            senderId: senderId,
            sender: sender,
            inputSender: inputSender,
        });
        this._client = client;
    }
}
exports.Forward = Forward;
