// telegram/client/bots.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.inlineQuery = void 0;
const tl_1 = require("../tl");
const inlineResults_1 = require("../tl/custom/inlineResults");
var GetInlineBotResults = tl_1.Api.messages.GetInlineBotResults;
// BotMethods
function inlineQuery(client, bot, query, entity, offset, geoPoint) {
    return __awaiter(this, void 0, void 0, function* () {
        bot = yield client.getInputEntity(bot);
        let peer = new tl_1.Api.InputPeerSelf();
        if (entity) {
            peer = yield client.getInputEntity(entity);
        }
        const result = yield client.invoke(new GetInlineBotResults({
            bot: bot,
            peer: peer,
            query: query,
            offset: offset || "",
            geoPoint: geoPoint,
        }));
        return new inlineResults_1.InlineResults(client, result, (entity = entity ? peer : undefined));
    });
}
exports.inlineQuery = inlineQuery;
